server.js

```
@@ -5,6 +5,101 @@ const app = express();
 
 app.use(cors());
 
+app.use(express.json());
+app.use(express.urlencoded({ extended: true }));
+
+//Create a new message
+app.post("/messages", (req, res) => {
+  let newPost = {
+    id: messages.length,
+    from: req.body.from,
+    text: req.body.text,
+    timeSent: dateTime,
+  };
+  if (newPost.from !== "" && newPost.text !== "") {
+    messages.push(newPost);
+    res.json(messages);
+  } else {
+    res.status(400).json(`Please fill all the fields`);
+  }
+});
+
+// Delete a message, by ID
+app.delete("/messages/delete/:id", (req, res) => {
+  let id = Number(req.params.id);
+  let index = messages.findIndex((message) => message.id === id);
+  if (index) {
+    messages.splice(index, 1);
+    res.json(`message with id: ${id} has been deleted`);
+  } else {
+    res.status(404).json("Please enter a valid ID to delete");
+  }
+});
+//Read all messages
+app.get("/messages", (req, res) => {
+  res.json(messages);
+});
+
+// add support for the client to be able to update a message's text or from property.
+app.put("/messages/:id", (req, res) => {
+  let id = Number(req.params.id);
+  let newText = req.body.text;
+  let found = messages.find((message) => message.id == id);
+  let updatedMsg = [];
+  messages.forEach((message) => {
+    if (message.id === found.id) {
+      message.text = newText;
+      updatedMsg.push(message);
+    } else {
+      updatedMsg.push(message);
+    }
+  });
+  res.json({ updatedMsg });
```

There are a few issues with this block. It is a bit overcomplicated.

1. You only have to update the text property on the `found` object if it is not `null`. As it is a reference to the object stored in the `messages` array and not a copy of that, meaning when you change it, you change the same thing that is part of the array too.
2. You don't have to create a copy of the `messages` array if it has nothing extra purposes as will be exactly the same as `messages` array.
3. Conventionally when a given entity has been updated, we only send the updated one back to the client and not everything, e.g `res.json(found);` so that client can see what has been the outcome of the update.
4. If the message has not been found, it should return an error as you did in the delete method